### PR TITLE
add mineOnlyExposedOres setting

### LINKED ISSUE #1168: Add "Ignore Not Exposed Ores" Option in AutoMine
## Describe your suggestion
This option should ignore ores that are not exposed to the air, because some servers use anti-Xray and they will replace not exposed stone with fake ores.

## Settings
There should be an option in automine that will automatically filter out not exposed ores.

## Context
For servers that use anti-Xray plugins, this would be very helpful to automine, because then we can only use legit mine, but we can also use Xray but only targeting exposed ores.

### PR DESCRIPTION
**Pull request by scorbett123**
adds an exposed Ores only setting.
if there is air no air on any side then it removes it from the goal.
fixes #1168

### CODE REVIEW

**Review by Enteerman — CHANGES_REQUESTED** [review]
LGTM, but why not have an integer setting since there are servers that hide ores except touches transparent block in a radius of two.

And, blocks like water, lava are see-through too and don't forget to include them.

**Comment by scorbett123** [comment]
converted to draft due to a bug in the latest commit.

File: src/api/java/baritone/api/Settings.java [hunk 1]
```diff
@@ -726,6 +726,18 @@
...
 
+    /**
+     * This will only allow baritone to mine exposed ores, can be used to stop ore obfuscators on servers that use them.
+     */
+    public final Setting<Boolean> allowOnlyExposedOres = new Setting<>(false);
+
+    /**
+     * When allowOnlyExposedOres is enabled this is the distance around to search.
+     * <p>
+     * I recommend keeping this value low as the amount of blocks that need to be scanned increases exponentially.
```
Comment (dominikaaaa) [inline src/api/java/baritone/api/Settings.java:737]: Rephrase this to be like
```diff
-      * I recommend keeping this value low as the amount of blocks that need to be scanned increases exponentially.
+      * It is recommended to keep this value low, as it exponentially increases calculation times.
```

File: src/main/java/baritone/process/MineProcess.java [hunk 2]
```diff
@@ -409,6 +410,8 @@ private void addNearby() {
                 // remove any that are implausible to mine (encased in bedrock, or touching lava)
                 .filter(pos -> MineProcess.plausibleToBreak(ctx, pos))
 
+                .filter(pos -> isNextToAir(ctx, pos))
```
Comment (dominikaaaa) [inline src/main/java/baritone/process/MineProcess.java:413]: Shouldn't this only calculate when the setting is enabled?

File: src/main/java/baritone/process/MineProcess.java [hunk 3]
```diff
@@ -452,4 +461,90 @@ public void mine(int quantity, BlockOptionalMetaLookup filter) {
...
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private static double lengthSq(double x, double y, double z) {
```
Comment (dominikaaaa) [inline src/main/java/baritone/process/MineProcess.java:539]: Can't you take a vec3d and square it?

File: src/main/java/baritone/process/MineProcess.java [hunk 4]
```diff
@@ -452,4 +461,90 @@ public void mine(int quantity, BlockOptionalMetaLookup filter) {
...
+        return (x * x) + (y * y) + (z * z);
+    }
+
+    public static boolean isTransparent(BlockPos pos, CalculationContext ctx) {
+        IBlockState blockState = ctx.bsi.get0(pos);
```
Comment (dominikaaaa) [inline src/main/java/baritone/process/MineProcess.java:544]: are you *sure* `get0` is the correct method here?

**Review by dominikaaaa — CHANGES_REQUESTED** [review]

File: src/main/java/baritone/process/MineProcess.java (hunk 4 above)
Comment (scorbett123) [reply src/main/java/baritone/process/MineProcess.java:544]: what do you suggest to use instead. If you use just get() then you get illegal state exceptions due to running it on the wrong thread. This is also the same as in the method above where it sees if it is plausible to break.

Comment (leijurv) [reply src/main/java/baritone/process/MineProcess.java:545]: I believe `get0` is correct in this case.

File: src/main/java/baritone/process/MineProcess.java [hunk 5]
```diff
@@ -452,4 +466,86 @@ public void mine(int quantity, BlockOptionalMetaLookup filter) {
...
+
+        return false;
+    }
+
+    public static boolean isTransparent(BlockPos pos, CalculationContext ctx) {
+        IBlockState blockState = ctx.bsi.get0(pos);
+        return blockState.getBlock() == Blocks.AIR ||
+                blockState.getBlock() == Blocks.FLOWING_LAVA ||
+                blockState.getBlock() == Blocks.FLOWING_WATER ||
+                blockState.getBlock() == Blocks.WATER;
```
Comment (leijurv) [inline src/main/java/baritone/process/MineProcess.java:549]: This helper `isTransparent` belongs in `MovementHelper`. I believe there may already be something equivalent there, but I'm not certain.

File: src/main/java/baritone/process/MineProcess.java [hunk 6]
```diff
@@ -452,4 +466,86 @@ public void mine(int quantity, BlockOptionalMetaLookup filter) {
             rescan(new ArrayList<>(), new CalculationContext(baritone));
         }
     }
+
+    public static boolean makeSphere(BlockPos pos, double radius, CalculationContext ctx) {
```
Comment (leijurv) [inline src/main/java/baritone/process/MineProcess.java:470]: This is extreme cringe. 

Literally everything about this function is cringe.

What is the use case for this? Ignore enteerman idk who that is.

Can't this be as simple as like:

```
for (int dx = -1; dx <= 1; dx++) {
    for (int dy = -1; dy <= 1; dy++) {
        for (int dz = -1; dz <= 1; dz++) {
            if (MovementHelper.isTransparent(pos.add(dx,dy,dz))) {
                return false;
            }
        }
    }
}
```

File: src/main/java/baritone/process/MineProcess.java (hunk 2 above)
Comment (leijurv) [reply src/main/java/baritone/process/MineProcess.java]: ya

File: src/main/java/baritone/process/MineProcess.java (hunk 6 above)
Comment (dominikaaaa) [reply src/main/java/baritone/process/MineProcess.java:470]: leij's example is how I was doing spheres and enteerman or someone made fun of it :clown:

Comment (leijurv) [reply src/main/java/baritone/process/MineProcess.java:470]: enteerman is the real clown then?

this isn't intended to be a sphere, why would it be?

Comment (dominikaaaa) [reply src/main/java/baritone/process/MineProcess.java:470]: # :joy_cat:

Comment (leijurv) [reply src/main/java/baritone/process/MineProcess.java:470]: yeah it shouldn't be a sphere

Comment (Enteerman) [reply src/main/java/baritone/process/MineProcess.java:470]: i am bad at english and sorry if i didnt explain it well. It wasn't intended to be a sphere. just on some servers that blocks that are two blocks from a transparent block are all legit. I think hypixel does that in UHC so you can check that out. You can ignore me, I don't care. I sketched this out and this is how is should work: https://imgur.com/0O7BkWt

Comment (Enteerman) [reply src/main/java/baritone/process/MineProcess.java:470]: I dont know why this method is large but just take the 6 blocks that are adjacent to the block, and take adjacents of the adjacents, to check if it is air.

Comment (Enteerman) [reply src/main/java/baritone/process/MineProcess.java:470]: Or use ur method idk i dont play the game anymore, ask someone else's opinion on this

Comment (ZacSharp) [reply src/main/java/baritone/process/MineProcess.java:470]: for me @leijurv's method looks *way* better but it only checks the direct and diagonal neighbours. I'd rather use
 ```
 for (int dx = -radius; dx <= radius; dx++) {
     for (int dy = -radius; dy <= radius; dy++) {
         for (int dz = -radius; dz <= radius; dz++) {
             if (dx + dy + dz <= radius && MovementHelper.isTransparent(pos.add(dx,dy,dz))) {
                 return true;
             }
         }
     }
 }
 ```
to check all blocks that have a manhattan distance of at most `radius` (I think that is was @Enteerman wants to show with his image)

**Comment by scorbett123** [comment]
sorry, accidentally pushed to this branch from the wrong repo, ignore latest commits.

File: src/api/java/baritone/api/Settings.java [hunk 7]
```diff
@@ -726,6 +726,18 @@
...
 
+    /**
+     * This will only allow baritone to mine exposed ores, can be used to stop ore obfuscators on servers that use them.
+     */
+    public final Setting<Boolean> allowOnlyExposedOres = new Setting<>(false);
+
+    /**
+     * When allowOnlyExposedOres is enabled this is the distance around to search.
+     * <p>
+     * It is recommended to keep this value low, as it exponentially increases calculation times and also to keep the
```
Comment (ZacSharp) [inline src/api/java/baritone/api/Settings.java:737]: this sentence is incomplete and the amount of checked blocks does not increase exponentially but by (approximately) `1.3*r**3 + 2*r**2 + 2.6*r + 1`

**Review by ZacSharp — CHANGES_REQUESTED** [review]

File: src/main/java/baritone/pathing/movement/MovementHelper.java [hunk 8]
```diff
@@ -562,4 +562,11 @@ static PlaceResult attemptToPlaceABlock(MovementState state, IBaritone baritone,
     enum PlaceResult {
         READY_TO_PLACE, ATTEMPTING, NO_OPTION;
     }
+    public static boolean isTransparent(Block b) {
+
+        return b== Blocks.AIR ||
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/MovementHelper.java:567]: space

File: src/main/java/baritone/process/MineProcess.java [hunk 9]
```diff
@@ -409,6 +410,13 @@ private void addNearby() {
                 // remove any that are implausible to mine (encased in bedrock, or touching lava)
                 .filter(pos -> MineProcess.plausibleToBreak(ctx, pos))
 
+                .filter(pos -> {
+                    if (Baritone.settings().allowOnlyExposedOres.value)
```
Comment (leijurv) [inline src/main/java/baritone/process/MineProcess.java:414]: plz use `{` in the if

File: src/main/java/baritone/process/MineProcess.java [hunk 10]
```diff
@@ -420,6 +428,22 @@ private void addNearby() {
         return locs;
     }
 
+    public static boolean isNextToAir(CalculationContext ctx, BlockPos pos) {
+        int radius = Baritone.settings().allowOnlyExposedOresDistance.value;
+        for (int dx = -radius; dx <= radius; dx++) {
+            for (int dy = -radius; dy <= radius; dy++) {
+                for (int dz = -radius; dz <= radius; dz++) {
+                    if (Math.abs(dx) + Math.abs(dy) + Math.abs(dz) <= radius
+                            && MovementHelper.isTransparent(ctx.getBlock(pos.getX()+dx, pos.getY()+dy, pos.getZ()+dz))) {
```
Comment (leijurv) [inline src/main/java/baritone/process/MineProcess.java:437]: run through autoformatter? put spaces before and after each `+`

**Review by leijurv — CHANGES_REQUESTED** [review]
LGTM, fix formatting to match baritone style and i'll merge

**Comment by leijurv** [comment]
@scorbett123 wait what, why did you rearrange those functions and classes? it makes the diff much bigger for no reason. can we leave the order the same and just add the new function? like how it was an hour ago?

**Comment by scorbett123** [comment]
Sorry, didn't mean to, something went wrong when committing, I think it might have been my auto formatter.

**Comment by leijurv** [comment]
ohh yea, that's true intellij has this `rearrange entries` option that i think you gotta untick

**Comment by scorbett123** [comment]
is this better?

**Comment by leijurv** [comment]
almost haha just one thing can you add curly braces to the `if`? 😹

**Comment by scorbett123** [comment]
Sorry, in my fourth attempt to fix the formatting and failing with using git I obviously forgot about those.
